sns_ik_lib/include/sns_ik/sns_vel_ik_base.hpp

```diff
@@ -0,0 +1,191 @@
+  typedef std::shared_ptr<SnsVelIkBase> Ptr;
+  typedef std::unique_ptr<SnsVelIkBase> uPtr;
+
+  enum class ExitCode {
+    Success,  // successfully solved the optimization problem
+    BadUserInput,  // user input failed basic checks (eg. inconsistent matrix size)
+    InfeasibleTask,  // there is no feasible solution to the primary task
+    InternalError  // there was an internal error in the solver (should never happen...)
+  };
+
+  /**
+   * Create a default solver with nJnt joints and no bounds on joint velocity
+   * @param nJnt: number of joints in the robot model (columns in the jacobian)
+   * @return: velocity solver iff successful, nullptr otherwise
+   */
+  static std::unique_ptr<SnsVelIkBase> create(int nJnt);
+
+  /**
+   * Create a default solver with constant bounds on the joint velocity
+   * @param dqLow: lower bound on the velocity of each joint
+   * @param dqUpp: upper bound on the velocity of each joint
+   * @return: velocity solver iff successful, nullptr otherwise
+   */
+  static std::unique_ptr<SnsVelIkBase> create(const Eigen::ArrayXd& dqLow,
+                                              const Eigen::ArrayXd& dqUpp);
+
+  // Make sure that class is cleaned-up correctly
+  virtual ~SnsVelIkBase() {};
+
+  /**
+   * Set the bounds on joint velocity.
+   * Set a bound to infinity (or an arbitrarily large value) to disable it.
+   * This method will update the number of joints in the solver
+   * Requirements: inputs must be the same size and dqLow < dqUpp
+   * @param dqLow: lower bound on the velocity of each joint
+   * @param dqUpp: upper bound on the velocity of each joint
+   * @return: true iff successful
+   */
+  bool setVelBnd(const Eigen::ArrayXd& dqLow, const Eigen::ArrayXd& dqUpp);
+
+  /**
+   * Solve a velocity IK problem with no null-space bias of joint-space optimization.
+   *
+   *  This method implements   "Algorithm 1: SNS algorithm"   from the paper:
+   *  "Control of Redundant Robots Under Hard Joint Constraint: Saturation in the Null Space"
+   *   by: Fabrizio Flacco, Alessandro De Luca, Oussama Khatib
+   *
+   * Solve for joint velocity dq and task scale s:
+   *
+   *  maximize: s
+   *  subject to:
+   *    s * dx = J * dq
+   *    0 < s <= 1
+   *    dqLow <= dq <= dqUpp      ( bounds set in constructor or setVelBnd() )
+   *
+   * @param J: Jacobian matrix, mapping from joint to task space. Size = [nTask, nJoint]
+   * @param dx: task velocity vector. Length = nTask
+   * @param[out] dq: joint velocity solution. Length = nJoint
+   * @param[out] taskScale: task scale.  fwdKin(dq) = taskScale*dx
+   *                            taskScale == 1.0  --> task was feasible
+   *                            taskScale < 1.0  --> task was infeasible and had to be scaled
+   * @return: Success: the algorithm worked correctly and satisfied the problem statement
+   *          otherwise: something went wrong
+   */
+  ExitCode solve(const Eigen::MatrixXd& J, const Eigen::VectorXd& dx,
```

Comment:
so this will be used like the following?
```
if( solver.solve(someJ, somedx) == sns_ik::SnsVelIkBase::ExitCode::Success ) ?
```
bit of a mouthful, but I can't think of a better way to do it :)

Reply:
That works, but I think that it is a bit cleaner to assign the exit code to a local variable. Then either use a conditional or a switch statement to decide what to do with the results. Take a look at the unit tests for a full example (there I use a simple conditional, as the check is simply pass / fail). Here is a simple example with a conditional check, assuming that `J` is the task jacobian, `dx` is the task goal velocity (twist), `dq` is the joint velocity, and `taskScale` is the task scale associated with the joint velocity solution.

    sns_ik::SnsVelIkBase::ExitCode exitCode = ikSolver->solve(J, dx, &dq, &taskScale);
    if (exitCode == sns_ik::SnsVelIkBase::ExitCode::Success) {
      // Code for success case
    } else {
      // code for the failure case
    }